### PR TITLE
Video Support in V1.0: part 1/3 : support matroskavideo

### PR DESCRIPTION
**Pull request by mohamedchebbii**
Hello,
I want to retsore the support of video files, I begin with matroskavideo and I will add tiff and asf format during next days.
Here a sample test results:
```
./bin/exiv2 -p x  ~/Downloads/sample_960x400_ocean_with_audio.mkv
Xmp.video.FileName                           XmpText    85  /home/mohamed.chebbi/Downloads/sample_960x400_ocean_with_audio.mkv
Xmp.video.FileSize                           XmpText     7  16.5413
Xmp.video.MimeType                           XmpText    14  video/matroska
Xmp.video.EBMLVersion                        XmpText     1  1
Xmp.video.EBMLReadVersion                    XmpText     1  1
Xmp.video.DocType                            XmpText     8  matroska
Xmp.video.DocTypeVersion                     XmpText     1  4
Xmp.video.DocTypeReadVersion                 XmpText     1  2
Xmp.video.TimecodeScale                      XmpText     5  0.001
Xmp.video.MuxingApp                          XmpText    13  Lavf58.45.100
Xmp.video.WritingApp                         XmpText    13  Lavf58.45.100
Xmp.video.Duration                           XmpText     5  46616
Xmp.video.TotalStream                        XmpText     1  1
Xmp.video.FrameRate                          XmpText     6  23.976
Xmp.video.Width                              XmpText     3  960
Xmp.video.Height                             XmpText     3  400
Xmp.video.AspectRatio                        XmpText     3  2.4
```

Please let me know If I miss anything, I will be happy to add It.
Best Regards,

### CODE REVIEW

**Comment by postscript-dev** [comment]
@mohamedchebbii
> Please let me know If I miss anything, I will be happy to add It.

Since the video support was removed, @neheb has continued using [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) to modernize the code to be compatible with C++17. I don't know which C++ version the video code currently supports but some small changes may be needed to comply. @neheb, do you have any thoughts on this?

@mohamedchebbii, please can you add at least one (small) test file to demonstrate that the code in your PR works. I don't know if we have a valid file in the `0.27-maintenance` branch, perhaps the code was untested by the CI? Also, I remember that Exiv2 discussed storing larger files outside of the Exiv2 repository (e.g., #896) and downloading them as needed, but I'm not sure what we decided.

**Comment by neheb** [comment]
I'd run through clang-tidy, yes. On a quick glance, I see constructors not marked as delete and marked public. I'm guessing this is copy/paste from somewhere.

File: include/exiv2/matroskavideo.hpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+namespace Exiv2 {
+
+// *****************************************************************************
+// class definitions
+
+    // Todo: Should be hidden
+    /*!
+      @brief Helper structure for the Matroska tags lookup table.
+     */
+    struct MatroskaTags {
```
Comment (neheb) [inline include/exiv2/matroskavideo.hpp:40]: A simple std::pair is good enough I think.

File: include/exiv2/matroskavideo.hpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+         */
+        void contentManagement(const MatroskaTags* mt, const byte* buf, long size);
+        /*!
+          @brief Calculates Aspect Ratio of a video, and stores it in the
+              respective XMP container.
+         */
+        void aspectRatio();
+
+    private:
+        //! @name NOT Implemented
```
Comment (neheb) [inline include/exiv2/matroskavideo.hpp:110]: Not implemented functions should be public and marked = delete;

File: include/exiv2/matroskavideo.hpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+          it will be deleted.
+     */
+    Image::UniquePtr newMkvInstance(BasicIo::UniquePtr io, bool create);
+
+    //! Check if the file iIo is a Matroska Video.
+    bool isMkvType(BasicIo& iIo, bool advance);
+
+}                                       // namespace Exiv2
+
+#endif                                  // #ifndef MATROSKAVIDEO_HPP_
```
Comment (neheb) [inline include/exiv2/matroskavideo.hpp:143]: No newline. Also why are these comments so far away?

File: src/matroskavideo.cpp [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+#include <cassert>
+#include<iostream>
+
+// *****************************************************************************
+// class member definitions
+namespace Exiv2 {
+    namespace Internal {
+
+    //! List of composite tags. They are skipped and the child tags are read immediately
+    uint64_t compositeTagsList[] = {
```
Comment (neheb) [inline src/matroskavideo.cpp:43]: Should be a constexpr std::array. Same with the below one.

File: src/matroskavideo.cpp [hunk 5]
```diff
@@ -0,0 +1,763 @@
...
+    /*!
+      Tag Look-up list for Matroska Type Video Files
+      The Tags have been categorized in 4 categories. Which are
+      mentioned as a comment in front of them.
+      s   --  Tag to be Skipped
+      sd  --  Tag to be Skipped along with its data
+      u   --  Tag used directly for storing metadata
+      ui  --  Tag used only internally
+     */
+    extern const MatroskaTags matroskaTags[] = {
```
Comment (neheb) [inline src/matroskavideo.cpp:77]: extern const is an old way of doing things. This should be constexpr and probably std::array.

File: include/exiv2/matroskavideo.hpp (hunk 1 above)
Comment (mohamedchebbii) [reply include/exiv2/matroskavideo.hpp]: Done https://github.com/Exiv2/exiv2/pull/2370/commits/65a01dc171f4dc230eed22754990d4c25dd0078d

File: include/exiv2/matroskavideo.hpp (hunk 2 above)
Comment (mohamedchebbii) [reply include/exiv2/matroskavideo.hpp]: done here https://github.com/Exiv2/exiv2/pull/2370/commits/65a01dc171f4dc230eed22754990d4c25dd0078d

File: include/exiv2/matroskavideo.hpp (hunk 3 above)
Comment (mohamedchebbii) [reply include/exiv2/matroskavideo.hpp]: done here https://github.com/Exiv2/exiv2/pull/2370/commits/65a01dc171f4dc230eed22754990d4c25dd0078d

File: src/matroskavideo.cpp (hunk 4 above)
Comment (mohamedchebbii) [reply src/matroskavideo.cpp]: done here https://github.com/Exiv2/exiv2/pull/2370/commits/65a01dc171f4dc230eed22754990d4c25dd0078d

File: src/matroskavideo.cpp (hunk 5 above)
Comment (mohamedchebbii) [reply src/matroskavideo.cpp]: done here https://github.com/Exiv2/exiv2/pull/2370/commits/65a01dc171f4dc230eed22754990d4c25dd0078d

Comment (mohamedchebbii) [reply src/matroskavideo.cpp]: Hello @postscript-dev, thank you for your comment, It's really helpful, I will try to add unit test these days  ( install all dependencies and resolve them take time more the development Itself :) )
Hello @neheb , thank you for your notes. I appreciate that, I fixed the code and pushed a new commit .
Gays, please bear with, I'm new in this project and I'm not aware It's history, if you have further advises , I'm happy to take them.
Regards

File: include/exiv2/matroskavideo.hpp (hunk 2 above)
Comment (neheb) [reply include/exiv2/matroskavideo.hpp]: hmm? I don't see = delete;

File: src/image.cpp [hunk 6]
```diff
@@ -100,6 +101,7 @@ constexpr auto registry = std::array{
     // needs to be before bmff because some ftyp files are handled as qt and
     // the rest should fall through to bmff
     Registry{ImageType::qtime, newQTimeInstance, isQTimeType, amRead, amNone, amRead, amNone},
+    Registry{ImageType::mkv,     newMkvInstance, isMkvType,   amRead, amNone, amRead, amNone},
```
Comment (benmccann) [inline src/image.cpp:104]: it looks like the format checker on the CI doesn't like the extra spaces
```suggestion
    Registry{ImageType::mkv, newMkvInstance, isMkvType, amRead, amNone, amRead, amNone},
```

File: include/exiv2/types.hpp [hunk 7]
```diff
@@ -14,6 +14,7 @@
 #include <limits>
 #include <sstream>
 #include <vector>
+#include <array>
```
[the remark this answers is not in the document]
Comment (mohamedchebbii) [reply include/exiv2/types.hpp]: Hello @benmccann,
thank you for your comment;
Yes If fact this CI test check the respect some formatting rules (to make to code prettier) witten in file .clan-format.
I fix them throw the command clang-format and pushed a new commit.

File: src/image.cpp (hunk 6 above)
Comment (mohamedchebbii) [reply src/image.cpp]: yes exactly

**Comment by kevinbackhouse** [comment]
Before this gets merged, I'd like to run the fuzzer on it for a few days. I'll wait until the rest of the code review is done, so please ping me when you think it's ready to merge.

**Comment by mohamedchebbii** [comment]
I'm sorry , I renamed my local branch and this PR was automatically close.
Please follow the PR :https://github.com/Exiv2/exiv2/pull/2373

File: include/exiv2/matroskavideo.hpp (hunk 2 above)
Comment (mohamedchebbii) [reply include/exiv2/matroskavideo.hpp]: my bad, It's here : 49f6acd

**Comment by benmccann** [comment]
All checks have passed on this PR. @neheb do you have any further comments or should @kevinbackhouse run the fuzzer now?

**Comment by neheb** [comment]
Needs rebasing.

**Comment by kevinbackhouse** [comment]
I started the fuzzer and it found a couple of crashes within a few minutes. The crash files are here: [crash.tar.gz](https://github.com/Exiv2/exiv2/files/9740266/crash.tar.gz)

Please could you read the README in the `fuzz` sub-directory and run the fuzzer yourself to check for other bugs?

**Comment by kevinbackhouse** [comment]
I notice that the only test in this PR is a very basic unit test. For a new source file like this, we need some proper test files in `test/data` with associated test scripts in `tests/`.

File: include/exiv2/matroskavideo.hpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        Calls contentManagement() or skips to next tag, if required.
+   */
+  void decodeBlock();
+  /*!
+    @brief Interpret tag information, and save it in the respective XMP container.
+    @param mt Pointer to current tag,
+    @param buf Pointer to the memory area with the tag information.
+    @param size Size of \em buf.
+   */
+  void contentManagement(const MatroskaTags* mt, const byte* buf, long size);
```
Comment (kevinbackhouse) [inline include/exiv2/matroskavideo.hpp:113]: Please don't use the `long` type. We're trying to eliminate it from the codebase. We usually use `size_t` for sizes.

Comment (mohamedchebbii) [reply include/exiv2/matroskavideo.hpp:113]: understood, thank you very much @kevinbackhouse  for your time.
I will analyse the crash report and run locally the fuzz test to fix them

**Comment by benmccann** [comment]
@mohamedchebbii this PR has a merge conflict that needs to be addressed

File: include/exiv2/matroskavideo.hpp (hunk 8 above)
Comment (mohamedchebbii) [reply include/exiv2/matroskavideo.hpp]: replacing long by size_t done here [f14bf95](https://github.com/Exiv2/exiv2/pull/2370/commits/f14bf95862ddaf4b02fafb5221b97319b4ef4e81)

**Comment by mohamedchebbii** [comment]
> Needs rebasing.

done here [f14bf95](https://github.com/Exiv2/exiv2/pull/2370/commits/f14bf95862ddaf4b02fafb5221b97319b4ef4e81)

**Pull request closed without merging**